### ISSUE TITLE
Assert does not work with ArraySegment

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#477]
Reported by @xplicit in nunit/nunitv2#39, moved here because we will not fix it in 2.6

This code throws an exception on .NET, while it should be passed. With mono 3.12 it works. The same issue with NUnit v 3.0

     Assert.AreEqual(new ArraySegment<byte>(),new ArraySegment<byte>());

**Comment by rprouse** [comment]
I tested this and it looks like a bug in the .NET Framework to me, so I am inclined to close this as invalid. It is crashing in NUnit in the call to GetEnumerator().

[According to the documentation](https://msdn.microsoft.com/en-us/library/4f47hffe(v=vs.110).aspx), `ArraySegment<T>` does not have a default constructor. All of it's constructors take an array of T and if you pass in null, it is supposed to throw an `ArgumentNullException`.

My thinking is that `new ArraySegment<byte>()` is invalid even if it works and that  the nunit framework should not be protecting the developer from everything they may do wrong.

I tested the more correct,

```C#
Assert.AreEqual(new ArraySegment<byte>(new byte[]{}), new ArraySegment<byte>(new byte[]{}));
```

and it works fine.

I am going to close this issue as invalid. @xplicit, if you feel I am doing so in error and you have a more complete use case or I have mis-interpreted this issue, I will be happy to re-open the issue and look at it again.

**Comment by xplicit** [comment]
I still think it's a bug in NUnit, not in .NET
1. ArraySegment does not have default constructor because it's a structure. All structures in .NET does not have default constructor, but this does not mean that you can't use them in `new` operator. For example: `DateTime dt = new DateTime();`

2. It's not always possible to call a constructor with empty byte array. For example, when use default constructor from generics.
  
    public T Create<T>()
    {
        return default(T);
    }

    Assert.AreEqual(Create<ArraySegment<byte>>(),Create<ArraySegment<byte>>());

2. .NET compares two newly created ArraySegments without an exception

    var a1 = new ArraySegment<byte>();
    var a2 = new ArraySegment<byte>();
    Console.WriteLine("{0}",a1==a2);  // true

3. MSTest performs `Assert.AreEqual(new ArraySegment<byte>(),new ArraySegment<byte>());` without an exception.

**Comment by CharliePoole** [comment]
I'm inclined to agree, since new ArraySegment<byte>().Equals(new ArraySegment<byte>()) returns true.

**Comment by CharliePoole** [comment]
The problem is that this falls into the IEnumerable comparison in NUnitEqualityComparer. It would be nicer if ArraySegment returned an empty enumerator when not based on an array, but it throws instead! The the trick would be to either detect an empty ArraySegment before asking for the enumerator or to catch the exception.

**Comment by rprouse** [comment]
Okay, I'm convinced, reopened :smile:

**Comment by CharliePoole** [comment]
Turns out to be a bit complicated, because the underlying implementation of ArraySegment differs from platform to platform! I have a fix that works when ArraySegment implements IEnumerable, but it actually doesn't do that on all of them.

That said, it's a useful issue in pulling out the need for some restructuring in NUnitEqualityComparer, even though it's not a high priority.

**Comment by CharliePoole** [comment]
ArraySegment doesn't implement IEnumerable in eitehr Silverlight or CF 3.5. It's not implemented in the version used by our portable build, probably because Silverlight is included in the mix. I'm going to set a flag if (new ArraySegment<int>() is IEnumerable) and use that to make the distinction.

**Comment by CharliePoole** [comment]
In the end, I decided it wasn't worth the effort for anything except the desktop .NET builds.

**Comment by rprouse** [comment]
This was closed by the PR. Do you consider it fixed 'enough' by the PR to close?

**Comment by xplicit** [comment]
I have a question: will last assert work? 

        var a1 = new ArraySegment<byte>();
        var a2 = new ArraySegment<byte>();
        Assert.IsTrue(a1==a2);   
        Assert.AreEqual(a1,a2);   
        Assert.IsTrue(a1==a2);   //will this work?

**Comment by rprouse** [comment]
The first and last assert look the same. Did you make a mistake typing or am I blind? :smile:

**Comment by xplicit** [comment]
Yes, they are the same. The question is about instance modification in Assert.AreEqual: are instances `a1` and `a2` modified or not.

**Comment by CharliePoole** [comment]
Since it's not Assert.AreEqual(ref a1, ref a2) it's not possible for them to be modified.

What gets modified in the code is the _copy_ of a1 and a2 that is being used by the constraint and saved in the result.

**Comment by CharliePoole** [comment]
@rprouse Yes,I think it's OK that it's closed.

I should have made a more extended comment: the reported issue is actually fixed. In working on the problem, I found another issue, which is not fixed: Assert.AreEqual won't work correctly using a runtime where ArraySegment does not implement IEnumerable. Currently, our tests return inconclusive results when this happens, which I admit is a bit of a cheat. But I didn't want to add tests when I don't even know how to make them pass! I guess I could add a new issue for this, although I think it's a design flaw in .NET.

One reason to work on this, in spite of it's clearly low importance: solving this problem means taking an object and somehow casting it to ArraySegment<sometype> where sometype is not known until runtime. If we could figure out how to do that, we could simplify our constraint code in many places.

BTW, the docs show that only 4.5 implements IEnumerable in ArraySegment. However, on my system, it works for 2.0 through 4.5. On Appveyor, it works for both 4.0 and 4.5, but it may not work on a machine with only 4.0. That's what led me to the dynamic check using Assume.That. Thanks to the mono guys, it works for all builds on Travis. If you search the web, you can find lots of complaints about how MS originally implemented ArraySegment. Apparently, it took till 4.5 to get it right.

Bottom line, due to how MS implements ArraySegment, tests don't always pass when they should, and sometimes fail catastrophically. We fixed the catastrophic failure and I doubt we can do more.

Fascinating!
